### ISSUE TITLE
Clang static analyzer identifies a potential NULL dereference

### ISSUE THREAD

**Issue by aaronmdjones** [troydhanson/uthash#111]
The following C99 code:

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "utlist.h"

struct entry
{
	struct entry    *prev;
	struct entry    *next;
	char             text[32];
};

struct entry *list = NULL;

static struct entry * __attribute__((malloc))
new_list_item(const char *const restrict text)
{
	struct entry *item = calloc(1, sizeof *item);

	if (! item)
		return NULL;

	(void) snprintf(item->text, sizeof item->text, "%s", text);

	DL_APPEND(list, item);
	return item;
}

int
main(void)
{
	if (! new_list_item("foo"))
		return EXIT_FAILURE;

	if (! new_list_item("bar"))
		return EXIT_FAILURE;

	if (! new_list_item("baz"))
		return EXIT_FAILURE;

	struct entry *item, *tmp;
	DL_FOREACH_SAFE(list, item, tmp)
	{
		DL_DELETE(list, item);
		free(item);
	}

	return EXIT_SUCCESS;
}
```

Results in the following Clang 3.8 static analyzer output:

```
$ /usr/bin/scan-build-3.8 -maxloop 32 \
  /usr/share/clang/scan-build-3.8/libexec/ccc-analyzer \
  -std=c99 -Wall -Wextra -Wpedantic test.c -o test

scan-build: Using '/usr/lib/llvm-3.8/bin/clang' for static analysis

test.c:45:3: warning: Access to field 'prev' results in a dereference
   of a null pointer (loaded from variable 'list')
                DL_DELETE(list, item);
                ^         ~~~~

./utlist.h:590:33: note: expanded from macro 'DL_DELETE'
#define DL_DELETE(head, del) DL_DELETE2(head, del, prev, next)
                                        ^          ~~~~

./utlist.h:450:15: note: expanded from macro 'DL_DELETE2'
                head->prev = del->prev;         \
                ~~~~       ^

1 warning generated.
scan-build: 1 bug found.
```

The code does not crash, and Valgrind memcheck does not complain about it, so it is probably a false positive, but you may want to look into it anyway. Perhaps the macros could be rewritten slightly to avoid the warning.

The following:

```
	struct entry *item, *tmp;
	DL_FOREACH_SAFE(list, item, tmp)
	{
		if (! list)
			break;

		DL_DELETE(list, item);
		free(item);
	}
```

Also silences the warning.

**Comment by Quuxplusone** [comment]
Hi Aaron, thanks for the report!  And especially thanks for the runnable test case; I still don't really understand how to use scan-build. :)

I've verified that scan-build produces that error.
I've verified that it's a false positive.
I've verified that we could silence it by adding `assert((head) != NULL);` to the top of DL_DELETE2.
The problem is, adding that assert silences the diagnostic only in debug mode. Passing -DNDEBUG on the scan-build command line causes the diagnostic to reappear.

When I build the unit tests with scan-build, I see a ton of errors, even without -DNDEBUG.
I've pasted an example at the bottom of this email.

On the one hand, it would be nice to have utlist be "scan-build clean", even if only in debug mode.
On the other hand, I have no way to test that on my local MacBook (and am about to lose access to the Linux box where I tested it just now), so it would be subject to bit-rot.

The other day I was actually thinking about ripping out all the asserts in "utlist.h", to eliminate a header dependency and make the macros speedier in debug mode. But if those asserts are actually functional (keeping the build clean), then maybe I shouldn't do that.

I'll add `assert((head) != NULL);` to the top of DL_DELETE2; but this won't make uthash or utarray "scan-build clean", and it might bit-rot over time.

```
$ /usr/bin/scan-build clang  -I../src -g -Wall -pedantic test40.c -o test40
scan-build: Using '/usr/bin/clang' for static analysis
test40.c:36:9: warning: Use of memory after it is freed
        HASH_DEL(users, s);
        ^~~~~~~~~~~~~~~~~~
../src/uthash.h:415:5: note: expanded from macro 'HASH_DEL'
    HASH_DELETE(hh,head,delptr)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/uthash.h:363:21: note: expanded from macro 'HASH_DELETE'
        uthash_free((head)->hh.tbl->buckets,                                     \
        ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/uthash.h:89:34: note: expanded from macro 'uthash_free'
#define uthash_free(ptr,sz) free(ptr)     /* free fcn                        */
                                 ^
test40.c:36:9: warning: Use of memory after it is freed
        HASH_DEL(users, s);
        ^~~~~~~~~~~~~~~~~~
../src/uthash.h:415:5: note: expanded from macro 'HASH_DEL'
    HASH_DELETE(hh,head,delptr)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/uthash.h:371:53: note: expanded from macro 'HASH_DELETE'
        if ((delptr) == ELMT_FROM_HH((head)->hh.tbl,(head)->hh.tbl->tail)) {     \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
../src/uthash.h:111:49: note: expanded from macro 'ELMT_FROM_HH'
#define ELMT_FROM_HH(tbl,hhp) ((void*)(((char*)(hhp)) - ((tbl)->hho)))
                                                ^
2 warnings generated.
scan-build: 1 bugs found.
```

**Comment by Quuxplusone** [comment]
Fixed (such as it is) by 693a09dbd9e22c534f03e8abc6a1bd8cdec12261.
